File: scandeps.ts
const CWL_IANA = "https://www.iana.org/assignments/media-types/application/cwl";

function scandeps_file_dir(
  base: string,
  doc: CWLObjectType,
  reffields: Set<string>,
  urlfields: Set<string>,
  loadref: (arg1: string, arg2: string) => CommentedMap | CommentedSeq | string | null,
  urljoin: (arg1: string, arg2: string) => string,
  nestdirs: boolean
): Array<CWLObjectType> {
  let r: Array<CWLObjectType> = [];
  let u = doc.get("location") || doc.get("path");
  if (u && !u.startsWith("_:")) {
    let deps: CWLObjectType = {
      "class": doc["class"],
      "location": urljoin(base, u)
    }
    if (doc["basename"]) {
      deps["basename"] = doc["basename"];
    }
    if (doc["class"] == "Directory" && doc["listing"]) {
      deps["listing"] = doc["listing"];
    }
    if (doc["class"] == "File" && doc["secondaryFiles"]) {
      deps["secondaryFiles"] = scandeps(
        base,
        doc["secondaryFiles"] as (CWLObjectType | Array<CWLObjectType>),
        reffields,
        urlfields,
        loadref,
        urljoin,
        nestdirs
      ) as CWLOutputAtomType;
    }
    if (nestdirs) {
      deps = nestdir(base, deps);
    }
    r.push(deps);
  } else {
    if (doc["class"] == "Directory" && doc["listing"]) {
      r = r.concat(scandeps(
        base,
        doc["listing"] as Array<CWLObjectType>,
        reffields,
        urlfields,
        loadref,
        urljoin,
        nestdirs
      ));
    } else if (doc["class"] == "File" && doc["secondaryFiles"]) {
      r = r.concat(scandeps(
        base,
        doc["secondaryFiles"] as Array<CWLObjectType>,
        reffields,
        urlfields,
        loadref,
        urljoin,
        nestdirs
      ));
    }
  }
  return r;
}
function scandeps_item(base: string,
                       doc: CWLObjectType,
                       reffields: Set<string>,
                       urlfields: Set<string>,
                       loadref: (param1: string, param2: string) => (CommentedMap | CommentedSeq | string | null),
                       urljoin: (param1: string, param2: string) => string,
                       nestdirs: boolean,
                       key: string,
                       v: any): Array<CWLObjectType> {
    let r: Array<CWLObjectType> = [];
    if (reffields.has(key)) {
        for (let u2 of aslist(v)) {
            if (u2 instanceof Map) {
                r.push(
                    ...scandeps(
                        base,
                        u2,
                        reffields,
                        urlfields,
                        loadref,
                        urljoin,
                        nestdirs,
                    )
                );
            } else {
                let subid = urljoin(base, u2);
                let basedf = new URL(base).hash;
                let subiddf = new URL(subid).hash;
                if (basedf == subiddf) {
                    continue;
                }
                let sub = loadref(base, u2);
                let deps2: CWLObjectType = {
                    "class": "File",
                    "location": subid,
                    "format": CWL_IANA,
                };
                let sf = scandeps(
                    subid,
                    sub,
                    reffields,
                    urlfields,
                    loadref,
                    urljoin,
                    nestdirs,
                );
                if (sf.length > 0) {
                    deps2["secondaryFiles"] = mergedirs(sf)
                }
                if (nestdirs) {
                    deps2 = nestdir(base, deps2);
                }
                r.push(deps2);
            }
        }
    } else if (urlfields.has(key) && key != "location") {
        for (let u3 of aslist(v)) {
            let deps = { "class": "File", "location": urljoin(base, u3) };
            if (nestdirs) {
                deps = nestdir(base, deps)
            }
            r.push(deps);
        }
    } else if ((doc.get("class") == "File" || doc.get("class") == "Directory") && (key == "listing" || key == "secondaryFiles")) {
        // should be handled earlier.
    } else {
        r.push(
            ...scandeps(
                base,
                v,
                reffields,
                urlfields,
                loadref,
                urljoin,
                nestdirs,
            )
        );
    }
    return r;
}
function scandeps(
    base: string,
    doc: CWLObjectType | CWLObjectType[],
    reffields: Set<string>,
    urlfields: Set<string>,
    loadref: (a: string, b: string) => CommentedMap | CommentedSeq | string | null,
    urljoin: (a: string, b: string) => string = urllib.parse.urljoin,
    nestdirs: boolean = true,
): CWLObjectType[] {
    let r: CWLObjectType[] = [];
    if (typeof doc === "object" && doc !== null && !Array.isArray(doc)) {
        if ("id" in doc) {
            if (typeof doc["id"] === 'string' && doc["id"].startsWith("file://")) {
                let df, _ = urllib.parse.urldefrag(doc["id"]);
                if (base !== df) {
                    r.push({ "class": "File", "location": df, "format": CWL_IANA });
                    base = df;
                }
            }
        }

        if (doc["class"] in ["File", "Directory"] && "location" in urlfields) {
            r = r.concat(scandeps_file_dir(base, doc, reffields, urlfields, loadref, urljoin, nestdirs));
        }

        for (let k in doc) {
            if (doc.hasOwnProperty(k)) {
                let v = doc[k];
                r = r.concat(scandeps_item(base, doc, reffields, urlfields, loadref, urljoin, nestdirs, k, v));
            }
        }
    } else if (Array.isArray(doc)) {
        for (let d of doc) {
            r = r.concat(
                scandeps(
                    base,
                    d,
                    reffields,
                    urlfields,
                    loadref,
                    urljoin,
                    nestdirs,
                )
            );
        }
    }

    if (r.length) {
        normalizeFilesDirs(r);
    }

    return r;
}

File: scandeps.py
CWL_IANA = "https://www.iana.org/assignments/media-types/application/cwl"

def scandeps_file_dir(base:str,
                      doc: CWLObjectType,
                      reffields: Set[str],
                      urlfields: Set[str],
                      loadref: Callable[[str, str], Union[CommentedMap, CommentedSeq, str, None]],
                      urljoin: Callable[[str, str], str],
                      nestdirs: bool) -> MutableSequence[CWLObjectType]:
    r: MutableSequence[CWLObjectType] = []
    u = cast(Optional[str], doc.get("location", doc.get("path")))
    if u and not u.startswith("_:"):
        deps: CWLObjectType = {
            "class": doc["class"],
            "location": urljoin(base, u),
        }
        if "basename" in doc:
            deps["basename"] = doc["basename"]
        if doc["class"] == "Directory" and "listing" in doc:
            deps["listing"] = doc["listing"]
        if doc["class"] == "File" and "secondaryFiles" in doc:
            deps["secondaryFiles"] = cast(
                CWLOutputAtomType,
                scandeps(
                    base,
                    cast(
                        Union[CWLObjectType, MutableSequence[CWLObjectType]],
                        doc["secondaryFiles"],
                    ),
                    reffields,
                    urlfields,
                    loadref,
                    urljoin=urljoin,
                    nestdirs=nestdirs,
                ),
            )
        if nestdirs:
            deps = nestdir(base, deps)
        r.append(deps)
    else:
        if doc["class"] == "Directory" and "listing" in doc:
            r.extend(
                scandeps(
                    base,
                    cast(MutableSequence[CWLObjectType], doc["listing"]),
                    reffields,
                    urlfields,
                    loadref,
                    urljoin=urljoin,
                    nestdirs=nestdirs,
                )
            )
        elif doc["class"] == "File" and "secondaryFiles" in doc:
            r.extend(
                scandeps(
                    base,
                    cast(MutableSequence[CWLObjectType], doc["secondaryFiles"]),
                    reffields,
                    urlfields,
                    loadref,
                    urljoin=urljoin,
                    nestdirs=nestdirs,
                )
            )
    return r
#------
def scandeps_item(base:str,
                      doc: CWLObjectType,
                      reffields: Set[str],
                      urlfields: Set[str],
                      loadref: Callable[[str, str], Union[CommentedMap, CommentedSeq, str, None]],
                      urljoin: Callable[[str, str], str],
                      nestdirs: bool,
                      key:str,
                      v:any) -> MutableSequence[CWLObjectType]:
    r: MutableSequence[CWLObjectType] = []
    if key in reffields:
        for u2 in aslist(v):
            if isinstance(u2, MutableMapping):
                r.extend(
                    scandeps(
                        base,
                        u2,
                        reffields,
                        urlfields,
                        loadref,
                        urljoin=urljoin,
                        nestdirs=nestdirs,
                    )
                )
            else:
                subid = urljoin(base, u2)
                basedf, _ = urllib.parse.urldefrag(base)
                subiddf, _ = urllib.parse.urldefrag(subid)
                if basedf == subiddf:
                    continue
                sub = cast(
                    Union[MutableSequence[CWLObjectType], CWLObjectType],
                    loadref(base, u2),
                )
                deps2: CWLObjectType = {
                    "class": "File",
                    "location": subid,
                    "format": CWL_IANA,
                }
                sf = scandeps(
                    subid,
                    sub,
                    reffields,
                    urlfields,
                    loadref,
                    urljoin=urljoin,
                    nestdirs=nestdirs,
                )
                if sf:
                    deps2["secondaryFiles"] = cast(
                        MutableSequence[CWLOutputAtomType], mergedirs(sf)
                    )
                if nestdirs:
                    deps2 = nestdir(base, deps2)
                r.append(deps2)
    elif key in urlfields and key != "location":
        for u3 in aslist(v):
            deps = {"class": "File", "location": urljoin(base, u3)}
            if nestdirs:
                deps = nestdir(base, deps)
            r.append(deps)
    elif doc.get("class") in ("File", "Directory") and key in (
        "listing",
        "secondaryFiles",
    ):
        # should be handled earlier.
        pass
    else:
        r.extend(
            scandeps(
                base,
                cast(Union[MutableSequence[CWLObjectType], CWLObjectType], v),
                reffields,
                urlfields,
                loadref,
                urljoin=urljoin,
                nestdirs=nestdirs,
            )
        )
    return r
#------
def scandeps(
    base: str,
    doc: Union[CWLObjectType, MutableSequence[CWLObjectType]],
    reffields: Set[str],
    urlfields: Set[str],
    loadref: Callable[[str, str], Union[CommentedMap, CommentedSeq, str, None]],
    urljoin: Callable[[str, str], str] = urllib.parse.urljoin,
    nestdirs: bool = True,
) -> MutableSequence[CWLObjectType]:
    """
    Search for external files references in a CWL document or input object.

    Looks for objects with 'class: File' or 'class: Directory' and
    adds them to the list of dependencies.

    :param base: the base URL for relative references.
    :param doc: a CWL document or input object
    :param urlfields: added as a File dependency
    :param reffields: field name like a workflow step 'run'; will be
      added as a dependency and also loaded (using the 'loadref'
      function) and recursively scanned for dependencies.  Those
      dependencies will be added as secondary files to the primary file.
    :param nestdirs: if true, create intermediate directory objects when
      a file is located in a subdirectory under the starting directory.
      This is so that if the dependencies are materialized, they will
      produce the same relative file system locations.
    :returns: A list of File or Directory dependencies
    """
    r: MutableSequence[CWLObjectType] = []
    if isinstance(doc, MutableMapping):
        if "id" in doc:
            if cast(str, doc["id"]).startswith("file://"):
                df, _ = urllib.parse.urldefrag(cast(str, doc["id"]))
                if base != df:
                    r.append({"class": "File", "location": df, "format": CWL_IANA})
                    base = df

        if doc.get("class") in ("File", "Directory") and "location" in urlfields:
            r.extend(scandeps_file_dir(base,doc,reffields,urlfields,loadref,urljoin,nestdirs))

        for k, v in doc.items():
            r.extend(scandeps_item(base,doc,reffields,urlfields,loadref,urljoin,nestdirs,k,v))
    elif isinstance(doc, MutableSequence):
        for d in doc:
            r.extend(
                scandeps(
                    base,
                    d,
                    reffields,
                    urlfields,
                    loadref,
                    urljoin=urljoin,
                    nestdirs=nestdirs,
                )
            )

    if r:
        normalizeFilesDirs(r)

    return r
